File: MoonEngine-Root/Assets/src/main/java/com/greentree/engine/moon/assets/serializator/IterableAssetSerializator.java
package com.greentree.engine.moon.assets.serializator;

import com.greentree.commons.util.classes.info.TypeInfo;
import com.greentree.commons.util.iterator.IteratorUtil;
import com.greentree.engine.moon.assets.key.AssetKey;
import com.greentree.engine.moon.assets.key.IterableAssetKey;
import com.greentree.engine.moon.assets.serializator.context.LoadContext;
import com.greentree.engine.moon.assets.serializator.manager.CanLoadAssetManager;
import com.greentree.engine.moon.assets.value.Value;
import com.greentree.engine.moon.assets.value.merge.MIValue;

public final class IterableAssetSerializator<T> extends AbstractIterableAssetSerializator<T> {
	
	public IterableAssetSerializator(Class<T> cls) {
		super(cls);
	}
	
	public IterableAssetSerializator(TypeInfo<T> type) {
		super(type);
	}
	
	public Value<Iterable<T>> load(LoadContext context, AssetKey ckey) {
		if(ckey instanceof IterableAssetKey key) {
			final var iter = IteratorUtil.clone(IteratorUtil.map(key, k-> {
				return context.load(ITER_TYPE, k);
			}));
			
			return new MIValue<>(iter);
		}
		return null;
	}
	
	@Override
	public boolean canLoad(CanLoadAssetManager manager, AssetKey key) {
		if(key instanceof IterableAssetKey k) {
			for(var s : k)
				if(!manager.canLoad(ITER_TYPE, s))
					return false;
			return true;
		}
		return false;
	}
	
}

File: MoonEngine-Root/Assets/src/main/java/com/greentree/engine/moon/assets/serializator/IterableAssetSerializator.kt
package com.greentree.engine.moon.assets.serializator

import com.greentree.commons.reflection.info.TypeInfo
import com.greentree.engine.moon.assets.asset.Asset
import com.greentree.engine.moon.assets.asset.MIAsset
import com.greentree.engine.moon.assets.key.AssetKey
import com.greentree.engine.moon.assets.key.IterableAssetKey
import com.greentree.engine.moon.assets.serializator.manager.AssetManager
import com.greentree.engine.moon.assets.serializator.marker.NotMyKeyType

class IterableAssetSerializator<T : Any> : AbstractIterableAssetSerializator<T> {
	constructor(cls: Class<T>) : super(cls)
	constructor(type: TypeInfo<T>) : super(type)

	override fun load(manager: AssetManager, key: AssetKey): Asset<Iterable<T>> {
		if(key is IterableAssetKey) {
			val iter = key.map { manager.load(ITER_TYPE, it) }
			return MIAsset(iter)
		}
		throw NotMyKeyType
	}
}